### DISCUSSION TITLE
Cannot perform %TypedArray%.prototype.set on a detached ArrayBuffer when using with uWebSockets

### DISCUSSION THREAD

**Discussion by tahayk** [socketio/socket.io#4281] [Q&A] [answered]
Hey
I'm trying to use uWebSockets with Socket.io, 
but I keep getting this error:

```
node:buffer:254
  TypedArrayPrototypeSet(target, source, targetStart);
  ^

TypeError: Cannot perform %TypedArray%.prototype.set on a detached ArrayBuffer
    at Buffer.set (<anonymous>)
    at _copyActual (node:buffer:254:3)
    at Function.concat (node:buffer:562:12)
    at onEnd (/Users/.../node_modules/engine.io/build/transports-uws/polling.js:126:32)
    at /Users/tahayk/.../node_modules/engine.io/build/transports-uws/polling.js:143:17

Node.js v17.0.1

```

Here is the example I used:

server.js:
```server.js
const { App } = require("uWebSockets.js");
const { Server } = require("socket.io");

const app = new App();
const io = new Server({ cors: {
        origin: "*",
        methods: ["GET", "POST", "PUT"]
    }});

io.attachApp(app);

io.on("connection", (socket) => {
    // ...
});

app.listen(9000, (token) => {
    if (!token) {
        console.warn("port already in use");
    }
});
```

demo.html
```demo.html
<!doctype html>
<html>
<head>
    <title>Socket.IO uWS</title>
</head>
<script src="node_modules/socket.io/client-dist/socket.io.js"></script>
<script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
<script>
    $(function () {
        var socket = io(
            'ws://127.0.0.1:9000',
            {
                // transports: ["websocket"] // ----> it works if you add this
            });

        socket.onAny((eventName, ...args) => {
            console.log("received event: " + eventName);
        });
    });
</script>
<body>
</body>
</html>
```

Any idea about this issue ? 
Thanks.

**Comment by darrachequesne** [comment]
Hi! I was not able to reproduce the issue with the code you provided. Could it be linked to the Node.js version (v17.0.1)? Could you please try with latest v16 for example?

**Comment by tahayk** [comment]
Hey, 

I tried with Node version `v16.0.0`; `v16.14.0`; `v17.5.0` and `v17.5.0` but still getting the same error.
PS: I updated the HTML snippet because I forgot to include jQuery.
This is my `package.json` file:
```json
{
  "name": "uws",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "start": "node server.js"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "socket.io": "^4.4.1",
    "uWebSockets.js": "github:uNetworking/uWebSockets.js#v20.6.0"
  }
}

```

**Comment by tahayk** [comment]
@darrachequesne 
Hey, 
I've just found out that the example I provided works fine on Google-Chrome, but I only get that error when I use Safari.

**Comment by e3dio** [comment]
@darrachequesne You are doing this wrong on this line https://github.com/socketio/engine.io/blob/a463d268ed90064e7863679bda423951de108c36/lib/transports-uws/polling.ts#L158 You are trying to concat Buffer views of arrayBuffers that are already gone. Allocate a buffer with correct size and copy arrayBuffers to it `Buffer.from(arrayBuffer).copy(allocatedBuffer, offset)` if the first arrayBuffer is not last

**Reply by e3dio** [reply]
The example file for onData https://github.com/uNetworking/uWebSockets.js/blob/master/examples/JsonPost.js#L63 is not the best because it concats arrayBuffers on each new chunk, this is a lot of extra copying because concat creates a new copy. Your method of gathering arrayBuffers into an array and then concat is slightly better except Buffer.from() creates a view not a copy, but even if you make copies that is still not efficient because there is 1 copy then final copy, also if 1st chunk is last you are making 2 copies when you need 0 copies. The best method is Buffer.allocUnsafe(size) and copy arrayBuffers to it, this gives 1 total copy per chunk, and if 1st chunk is last you skip that and use the 1st arrayBuffer for zero copies. You get allocated Buffer size from req.getHeader('content-length') then confirm total size matches and is under limit

**Reply by darrachequesne** [reply]
That makes sense. Could you please open a PR for this?

**Reply by e3dio** [reply]
Added PR here for review https://github.com/socketio/engine.io/pull/642

**Comment by tahayk** [comment]
@darrachequesne any update on this ? thanks.

**Reply by darrachequesne** [reply]
I'm looking into it.

**Comment by darrachequesne** [comment, marked as answer]
This should be fixed by https://github.com/socketio/engine.io/commit/33674403084c329dc6ad026c4122333a6f8a9992, included in `engine.io@6.1.3`.

In Safari, the server did receive two chunks during the Socket.IO handshake, the first being empty:

```
<Buffer > isLast:false
<Buffer 34 30> isLast:true
```
